PostgreSQL Network Disruption and Pod Delete Chaos Workflow (#8)
chaoscarnival/bootcamps: PostgreSQL/README.md

# PostgreSQL Network Disruption and Pod Delete Chaos Workflow

### PreRequisites 

#### We are using Zalando Operator for demo. If you have already postgres installed you can skip below steps

- Install the [Configuration manifest](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/configuration.yaml) on Kubernetes
- Install the [Operator manifest](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/operator.yaml) on Kubernetes

> - Install the `Postgres Cluster` on Kubernetes with the `Prometheus Exporter` configured. 

Reviewer: Mention a note that we have to install any one of the below Postgres Cluster

  - Single node [Postgres Cluster](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/weak-postgres-manifest.yaml) 

>   - As a 2 node master - slave [Postgres Cluster](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/resilient-postgres-manifest.yaml)  

Reviewer: Node node(master & slave) Postgres Cluster

> - Install the Litmus (Portal) Control Plane. You can use the step provided in the [Portal Documentation](https://github.com/litmuschaos/litmus/tree/master/litmus-portal#applying-k8s-manifest)

Reviewer: portal -- ChaosCenter

> - Run `echo $(kubectl get secret zalando.postgres-application.credentials -n postgres -o 'jsonpath={.data.password}')`

Reviewer: Can we mention the step where we need this password?

**Note:** Prefered `postgres` namespace for postgres cluster and `monitoring` namespace for prometheus and grafana.

> #### If you have already postgres cluster 

Reviewer: we can add step#<number> in the begg of these steps

- Update host, dbname, username, password and port in [secret.yaml](https://github.com/litmuschaos/test-tools/blob/master/custom/workflow-helper/postgres-helper/application/secret.yaml) file and create it in control plane or agent namespace.

> - Install postgres load generator for read and write operations

Reviewer: mentioned that we have to install any of the below load generators

  - Using custom load generator [Deployment manifest](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/deploy-load.yaml)
  - Using Jmeter load generator [Deployment manifest](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/deploy-jmeter.yaml)

> - Setup the Prometheus monitoring infra via the [Prometheus Operator](https://github.com/litmuschaos/test-tools/tree/master/custom/workflow-helper/postgres-helper/monitoring/prometheus) with suitable [Configuration](https://github.com/litmuschaos/test-tools/blob/master/custom/workflow-helper/postgres-helper/monitoring/prometheus/02-prometheus-configMap.yaml)

Reviewer: change the suitable configuration to a more meaningful statement

> - Apply the [Grafana](https://github.com/litmuschaos/test-tools/tree/master/custom/workflow-helper/postgres-helper/monitoring/k8s-grafana) and use [Dashboard](https://github.com/litmuschaos/test-tools/blob/master/custom/workflow-helper/postgres-helper/monitoring/k8s-grafana/dashboard.json) to moniter.

Reviewer: Install the Grafana

> *Please retain only the prometheus scrape configuration needed for this experiment (i.e., chaos exporter). These link shared above contains additional service monitors*

Reviewer: service monitors --> jobs

## Scenarios :

### Weak Scenario:
![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/weak.png)
![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/weak1.png)

In a single replica scenario(weak) after getting terminated the postgres container so db will be unavailable until it restarts again. So users will be unable to read/write queries.  

### Resilient Scenario:
![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/resilient.png)
![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/resilient1.png)

In 2 replica scenarios (resilient) after getting terminated the master, slave will be available. So users will be able to read/write queries.

### Demo Playbook

- Describe the Chaos Workflow manifests describing how network loss and pod delete are injected on master node & the cmdProbe and promProbe to verify application state. 

> 

Reviewer: rename the ss files to a meaningful name

# PostgreSQL Network Disruption and Pod Delete Chaos Workflow

## PreRequisites 

> ## Step 1. Install postgres using Zalando Operator for demo. You can skip 1st step if postgres is already installed 

Reviewer: ```suggestion
## Step 1. Install Postgres using Zalando Operator. You can skip it if Postgres is already installed 
```

- Install the [Configuration manifest](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/configuration.yaml) on Kubernetes
- Install the [Operator manifest](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/operator.yaml) on Kubernetes

> - Install the `Postgres Cluster` on Kubernetes with the `Prometheus Exporter` as sidecars. Install any one of the below Postgres cluster.

Reviewer: ```suggestion
- Install the `Postgres Cluster`  along with the `Prometheus Exporter` sidecar in the Kubernetes cluster. Install any one of the below Postgres cluster.
```

  - Single node(only master) [Postgres Cluster](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/weak-postgres-manifest.yaml) 
  - Two node(master & slave) [Postgres Cluster](https://raw.githubusercontent.com/litmuschaos/test-tools/master/custom/workflow-helper/postgres-helper/application/resilient-postgres-manifest.yaml)  

> - Install the Litmus(ChaosCenter). You can use the step provided in the [Portal Documentation](https://github.com/litmuschaos/litmus/tree/master/litmus-portal#applying-k8s-manifest)

Reviewer: ```suggestion
- Install the Litmus(ChaosCenter). Follow the step provided in the [Portal Documentation](https://github.com/litmuschaos/litmus/tree/master/litmus-portal#applying-k8s-manifest)
```

> - Get the password by executing the given command `echo $(kubectl get secret zalando.postgres-application.credentials -n postgres -o 'jsonpath={.data.password}')`. Need password to run cmdProbe inside workflow and load-generator. Update this password in 2nd step `secret.yaml` file.

Reviewer: ```suggestion
- Get the password by executing the given command `echo $(kubectl get secret zalando.postgres-application.credentials -n postgres -o 'jsonpath={.data.password}')`. It is required  to run cmdProbe inside workflow and to generate the load(via load-generator). It should be provided inside the `secret.yaml` file(will see in next/2nd step)
```

**Note:** Prefered `postgres` namespace for postgres cluster and `monitoring` namespace for prometheus and grafana.

> ## Step 2. Load and Monitoring Installation

Reviewer: ```suggestion
## Step 2. Install Load Generator And Monitoring Components
```

- Update host, dbname, username, password and port in [secret.yaml](https://github.com/litmuschaos/test-tools/blob/master/custom/workflow-helper/postgres-helper/application/secret.yaml) file. 
  
  **Note**: 
    - Values can be taken from application secret/config file and Apply it in litmus agent namespace. 

>     - Update in encoded (Base64-encoded DER format). Example: `echo 'postgres' | base64`, Output is `cG9zdGdyZXMK`, update this encoded value in `secret.yaml` manifest. 

Reviewer: ```suggestion
    - The values should be provided in encoded (Base64-encoded DER) format.
          Example: `echo 'postgres' | base64`
          Output is `cG9zdGdyZXMK`
   update the encoded value in `secret.yaml` manifest. 
```

>     - If zalando operator is installed then only need to update password other values are default.

Reviewer: ```suggestion
    - If the zalando operator was used to install postgres then update only the password file, other values should remain the same(default).
```

- Install the [Grafana](https://github.com/litmuschaos/test-tools/tree/master/custom/workflow-helper/postgres-helper/monitoring/k8s-grafana) and use [Dashboard](https://github.com/litmuschaos/test-tools/blob/master/custom/workflow-helper/postgres-helper/monitoring/k8s-grafana/dashboard.json) to moniter.

## UseCase & Hypothesis

- Network loss and pod-delete chaos is injected on postgresql master pod. 
- Network loss cause to application master node by injecting packet loss using litmus
- Pod Delete cause (forced/graceful) failure of master pod of an application. Tests deployment sanity (node availability & uninterrupted service) and recovery workflow of the application
- The application pod should be healthy once chaos is completed. Service-requests should be served despite chaos.

## Scenarios :

### Single node(master) Scenario:
- Application installed with single node(master) as a statefulset.

> - Experiment Network-Loss: Due to network loss accessibility to endpoint disrupts and unable to read/write.

Reviewer: ```suggestion
- Experiment Network-Loss: It disrupts the accessibility of postgres endpoints that blocks the read and write operations.
```

> - Experiment Pos-Delete: Once master pod is terminated the database will be inaccessible until it restarts again, mean while unable to read/write queries. 

Reviewer: ```suggestion
- Experiment Pod-Delete: Once master pod is terminated the database will be inaccessible until it restarts again, mean while unable to read/write queries. 
```

> - We can see in first chaos red dotted area for network loss and 2nd for pod-delete. 

Reviewer: ```suggestion
- We can visualize the network loss with the first red dotted area and pod-delete by the second red dotted area. 
```

![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/weak-resilient-insertion-rate-dashboard.png)
![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/weak-insertion-rate-and-deletion-rate-dashboard.png)

- Experiment logs and chaosresult has been added [here](https://github.com/oumkale/bootcamps/tree/postgres/PostgreSQL/result/weak).

### Two node(master-slave) Scenario:

- Application installed with two node(master-slave) as a statefulset.
- Experiment Network-Loss: Due to network loss master is inaccessible, after down time slave will become master and master will become slave. Due to database accessible Read-Write requests get successed. postgres operator carries out the rolling update, it re-spawns pods of each managed StatefulSet one-by-one with the master slave.

> - Experiment Pos-Delete: Once master pod is terminated the slave will become immediately master and available. Due to database accessiblity Read-Write requests get successed.

Reviewer: ```suggestion
- Experiment Pod-Delete: Once the master pod is terminated the slave will become master immediately and keep available. Due to database accessibility Read-Write requests get succeed.
```

![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/resilient-insertion-rate-dashboard.png)
![image](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/images/resilient-insertion-rate-and-deletion-rate-dashboard.png)

- Experiment logs and chaosresult has been added [here](https://github.com/oumkale/bootcamps/tree/postgres/PostgreSQL/result/resilient).

> **Note** : In a two-node(master-slave) cluster during network loss, there is downtime from Postgres. This is because it takes a certain time to check accessibility with current master role and once it achive max time limit, Immediately transfer the master role to the slave during the chaos. In master, pod delete there is no other way than allowing master role to slave since master is deleted.

Reviewer: ```suggestion
**Note**: In a two-node(master-slave) cluster during network loss, there is downtime from Postgres. This is because it takes a certain amount of time to check accessibility with the current master role and once it achieves the max time limit, Immediately transfer the master role to the slave during the chaos. In master, pod delete there is no other way than allowing the master role to slave since the master is deleted.
```

### Workflow Details

- PostgreSQL Network Disruption and Pod Delete Chaos workflow:
  - [workflow](https://github.com/oumkale/bootcamps/blob/postgres/PostgreSQL/workflow/workflow-prom.yaml) using CMDProbe for postgres database check and promProbe for prometheus query benchmark check.

- Workflow use case. 
  - CMDProbe for postgres database check plays an important role for identifying the error during chaos injection. We can see both weak and resilient experiment logs to see the error.
  - PromProbe is getting used to check the accessibility of postgresql state and benchmark check for prometheus metrics on a given query.

> **Note**: Make sure your runtime is docker, containerd or crio. Default is docker. 

Reviewer: ```suggestion
**Note**: Make sure your runtime should be one of the docker, containerd, or crio. Default is docker. 
```

>   - If it is containerd then update chaosengine with`CONTAINER_RUNTIME` and `SOCKET_PATH` with `containerd` and `/run/containerd/containerd.sock` in workflow manifest.

Reviewer: ```suggestion
  - If it is containerd then update chaosengine with`CONTAINER_RUNTIME` and `SOCKET_PATH` with `containerd` and `/run/containerd/containerd.sock` respectively in workflow manifest.
```
